app/Models/ScheduledRepayment.php

```diff
@@ -5,10 +5,45 @@
 use Illuminate\Database\Eloquent\Factories\HasFactory;
 use Illuminate\Database\Eloquent\Model;
 use Illuminate\Database\Eloquent\Relations\BelongsTo;
+use Illuminate\Notifications\Notifiable;
 
 class ScheduledRepayment extends Model
 {
-    use HasFactory;
+    use HasFactory, Notifiable;
+
+    public function repaid() {
+        $this->outstanding_amount = 0;
+        $this->status = ScheduledRepayment::STATUS_REPAID;
+    }
+
+    public function updateRepaid() {
+        if ($this->outstanding_amount == 0 || $this->status == ScheduledRepayment::STATUS_REPAID) {
+            $this->repaid();
+        }
+    }
+
+    protected static function booted()
+    {
+        static::creating(function (ScheduledRepayment $scheduledRepayment) {
+            $scheduledRepayment->outstanding_amount =
+                $scheduledRepayment->outstanding_amount ?? $scheduledRepayment->amount;
+            $scheduledRepayment->updateRepaid();
+            if ($scheduledRepayment->status != ScheduledRepayment::STATUS_DUE) {
+                $scheduledRepayment->load('loan');
+                $scheduledRepayment->loan->refresh();
+                $scheduledRepayment->loan->outstanding_amount -=
+                    $scheduledRepayment->amount - $scheduledRepayment->outstanding_amount;
+                $scheduledRepayment->loan->save();
+            }
+        });
+        static::updating(function (ScheduledRepayment $scheduledRepayment) {
+            $scheduledRepayment->updateRepaid();
+            if ($scheduledRepayment->outstanding_amount < $scheduledRepayment->amount &&
+                $scheduledRepayment->outstanding_amount > 0) {
+                $scheduledRepayment->status = ScheduledRepayment::STATUS_PARTIAL;
+            }
+        });
```

Comment:
The test cases seem to imply that the creation of a `ScheduledRepayment` with `STATUS_REPAID` status should automatically deduct the amount from its associated `Loan`. Is this possible without model events?

https://github.com/SuibianP/laravel-code-challenge-2/blob/a12e497dd095a696bf4cd29d3c8579aa40e91559/tests/Unit/LoanServiceTest.php#L153-L203

Reply:
Also, since we want to always maintain the invariant, I guess it makes sense to check for that at the model level. Otherwise we may produce the same logic somewhere else in the codebase. 🤔

Reply:
The test is calling `LoadService@repayLoan` so we totally can move these logic to the service.
The model observers & events are really cool but it will not a best choice when we want to implement a complex logic.
It will become really hard to debug because the events are dispatching in background and can be missed in the main flow when we browsing the code.

Reply:
In that way, the outstanding amount will be 5000-1668 not 0 at the end 🤔

Reply:
Some thoughts arising from this question,

For database, do we prefer calculation or size? For example, status can be calculated from outstanding_amount and amount. Is it redundant?

Furthermore, is there some other better way to maintain the invariant? For example, a better way to keep the loan outstanding amount equal to the sum of all associated outstanding repayments.